File: backend/GoIsland.Api/Services/Auth/AuthService.cs
using GoIsland.Api.DTOs.Auth;
using GoIsland.Api.Models;
using GoIsland.Api.Repositories;
using GoIsland.Api.Services.Security;

namespace GoIsland.Api.Services.Auth;

public class AuthService : IAuthService
{
    private readonly IUserRepository _users;
    private readonly IPasswordHasher _passwordHasher;
    private readonly IJwtTokenService _jwtTokenService;

    public AuthService(IUserRepository users, IPasswordHasher passwordHasher, IJwtTokenService jwtTokenService)
    {
        _users = users;
        _passwordHasher = passwordHasher;
        _jwtTokenService = jwtTokenService;
    }

    public async Task<AuthResponse?> RegisterAsync(RegisterRequest request)
    {
        var email = request.Email.Trim().ToLowerInvariant();
        if (await _users.GetByEmailAsync(email) is not null)
        {
            return null;
        }

        var role = string.IsNullOrWhiteSpace(request.Role) ? UserRoles.Tourist : request.Role.Trim();
        if (!UserRoles.PublicRegistration.Contains(role))
        {
            role = UserRoles.Tourist;
        }

        var user = new User
        {
            FullName = request.FullName.Trim(),
            Email = email,
            PasswordHash = _passwordHasher.Hash(request.Password),
            Role = role
        };

        await _users.AddAsync(user);
        return CreateAuthResponse(user);
    }

    public async Task<AuthResponse?> LoginAsync(LoginRequest request)
    {
        var user = await _users.GetByEmailAsync(request.Email);
        if (user is null || !_passwordHasher.Verify(request.Password, user.PasswordHash))
        {
            return null;
        }

        return CreateAuthResponse(user);
    }

    private AuthResponse CreateAuthResponse(User user)
    {
        var token = _jwtTokenService.CreateToken(user);

        return new AuthResponse
        {
            Token = token.Token,
            ExpiresAt = token.ExpiresAt,
            User = ToResponse(user)
        };
    }

    public static UserResponse ToResponse(User user)
    {
        return new UserResponse
        {
            Id = user.Id,
            FullName = user.FullName,
            Email = user.Email,
            Role = user.Role,
            CreatedAt = user.CreatedAt
        };
    }
}

File: frontend/src/services/authService.ts
import { api } from './api';
import type { AuthResponse, LoginRequest, RegisterRequest, UserResponse } from '../types';

export const authService = {
  login: async (data: LoginRequest): Promise<AuthResponse> => {
    const response = await api.post<AuthResponse>('/auth/login', data);
    return response.data;
  },

  register: async (data: RegisterRequest): Promise<AuthResponse> => {
    const response = await api.post<AuthResponse>('/auth/register', data);
    return response.data;
  },

  getMe: async (): Promise<UserResponse> => {
    const response = await api.get<UserResponse>('/auth/me');
    return response.data;
  },

  updateProfile: async (fullName: string): Promise<UserResponse> => {
    const response = await api.put<UserResponse>('/users/profile', { fullName });
    return response.data;
  },
};
export type { LoginRequest, RegisterRequest };
